js/src/chat/chat-tools.ts

```diff
@@ -0,0 +1,296 @@
+  hidden = false
+
+  constructor() {
+    super()
+    this.titleTemplate = "Running {title}"
+    this.icon = '<div class="spinner-border" role="status"></div>'
+  }
+
+  connectedCallback() {
+    super.connectedCallback()
+    window.addEventListener("shiny-tool-request-hide", this.#onToolRequestHide)
+  }
+
+  disconnectedCallback() {
+    super.disconnectedCallback()
+    window.removeEventListener(
+      "shiny-tool-request-hide",
+      this.#onToolRequestHide,
+    )
+  }
+
+  #onToolRequestHide = (event: CustomEvent) => {
+    if (event.detail.request_id === this.requestId) {
+      this.hidden = true
+    }
+  }
+
+  render() {
+    if (this.hidden) {
+      return html``
+    }
+
+    const bodyContent = html`
+      <shiny-markdown-stream
+        content="${markdownCodeBlock(this.arguments, "json")}"
+        content-type="markdown"
+        ?streaming=${false}
+      ></shiny-markdown-stream>
+    `
+
+    return this.renderCard(
+      html`<div class="shiny-tool-request__arguments">
+        <strong>Tool arguments</strong> ${bodyContent}
+      </div>`,
+    )
+  }
+}
+
+/**
+ * Web component that displays the result of a tool execution.
+ *
+ * @element shiny-tool-result
+ */
+export class ShinyToolResult extends ShinyToolCard {
+  @property({ type: String, attribute: "request-call" })
+  requestCall?: string
+
+  @property({ type: String })
+  status!: string
+
+  @property({ type: Boolean, attribute: "show-request" })
+  showRequest = true
+
+  @property({ type: String })
+  value!: string
+
+  @property({ type: String, attribute: "value-type" })
+  valueType!: string
+
+  constructor() {
+    super()
+    this.titleTemplate = "{title}"
+  }
+
+  connectedCallback() {
+    super.connectedCallback()
+    // Set status class and icon based on status
+    if (this.status === "error") {
+      this.classStatus = "text-danger"
+      this.icon = ICONS.exclamationCircleFill
+      this.titleTemplate = "{title} failed"
+    } else if (!this.icon) {
+      this.icon = ICONS.wrenchAdjustable
+    }
+    // Emit event to hide the corresponding tool request
+    this.dispatchEvent(
+      new CustomEvent("shiny-tool-request-hide", {
+        detail: { request_id: this.requestId },
+        bubbles: true,
+        cancelable: true,
+      }),
+    )
+  }
+
+  #renderResult() {
+    let result: string | TemplateResult = ""
+
+    if (this.valueType === "html") {
+      result = html`${unsafeHTML(this.value)}`
+    } else if (this.valueType === "text") {
+      result = html`<p>${this.value}</p>`
+    } else {
+      // markdown, code, or default
+      if (this.valueType !== "markdown") {
+        // If value_type is "code", we format it as a markdown code block
+        result = markdownCodeBlock(this.value, "text")
+      }
+
+      result = html`<shiny-markdown-stream
+        content=${result || this.value}
+        content-type="markdown"
+        ?streaming=${false}
+      ></shiny-markdown-stream>`
```

Comment:
Similar point to https://github.com/posit-dev/shinychat/pull/52/files#r2246381129